### DISCUSSION TITLE
Problems with polygon equality upon rotation

### DISCUSSION THREAD

**Discussion by dhananjaisharma10** [shapely/shapely#1773] [Q&A] [answered]
## Expected behavior and actual behavior.

A polygon rotated by angles `theta`, `theta + 180`, `theta - 360`, `theta - 180` should lead to the same polygon. But it does not. Why?

## Steps to reproduce the problem.

```python
def test():
    from shapely.geometry import Polygon
    from shapely.affinity import rotate

    original_polygon = Polygon([(15, 33), (25, 33), (25, 17), (15, 17)])
    # Rotate original_polygon CCW by 30 degrees.
    rot_polygon = rotate(original_polygon, angle=30, origin="center", use_radians=False)

    for angle in [210, -150, -330]:
        # Rotate original_polygon CCW by angle
        double_rot_polygon = rotate(original_polygon, angle=angle, origin="center", use_radians=False)
        print(f"{'Equal' if double_rot_polygon.equals(rot_polygon) else 'Different'}")


if __name__ == '__main__':
    test()
```

Output:
```bash
Different
Different
Different
```

## Operating system

Ubuntu 20.04

## Shapely version and provenance

Version `1.8.1` installed from PyPI using pip.

**Comment by jorisvandenbossche** [comment, marked as answer]
There seem to be tiny floating point differences between both. Using `double_rot_polygon` calculated with `angle = 210`:

```python
>>> print(double_rot_polygon.normalize())
POLYGON ((11.669872981077804 29.428203230275503, 20.33012701892219 34.4282032302755, 28.33012701892219 20.571796769724486, 19.669872981077805 15.571796769724486, 11.669872981077804 29.428203230275503))
>>> print(rot_polygon.normalize())
POLYGON ((11.669872981077805 29.428203230275507, 20.330127018922195 34.42820323027551, 28.330127018922195 20.571796769724486, 19.669872981077805 15.571796769724491, 11.669872981077805 29.428203230275507))

>>> rot_polygon.normalize().equals_exact(double_rot_polygon.normalize(), tolerance=1e-5)
True
```

Using the `equals` method, such small differences do matter. Using `equals_exact`, you can specify a tolerance for comparing coordinates (this function expects the coordinates in the exact same order, that's the reason I am calling `normalize()` first).

See https://shapely.readthedocs.io/en/stable/reference/shapely.equals_exact.html

**Reply by dhananjaisharma10** [reply]
Thank you. In theory, that difference should not exist, I believe.
Also, would it be a good idea to have a function that combines `equals_exact` and `normalize` together?

**Reply by jorisvandenbossche** [reply]
> In theory, that difference should not exist, I believe.

I think that is unavoidable when working with floating point values. 
Example part of the calculation with an angle of 30 and 210:

```
In [2]: import math

In [3]: angle = 30 * math.pi / 180.0

In [4]: math.cos(angle) * 15 + (-math.sin(angle) * 33) + 0.0
Out[4]: -3.5096189432334164

In [5]: angle = 210 * math.pi / 180.0

In [6]: math.cos(angle) * 15 + (-math.sin(angle) * 33) + 0.0
Out[6]: 3.5096189432334253
```

Just this part with some (co)sinus calculations gives a slightly different result.

> Thank you. In theory, that difference should not exist, I believe. Also, would it be a good idea to have a function that combines `equals_exact` and `normalize` together?

There is an open PR to add a `normalize` keyword to `equals_exact`, making this a bit easier
